app/vlselect/logsql: add offset query option

Fusl opened the pull request:
* offset the tailing window to be at least 1 second behind real-time
* allow adjusting the offset to an arbitrary value to fit custom use-cases where ingestion of logs is randomized or delayed
* limit the length of the time window to prevent out of memory situations
* use `slices.SortStableFunc` over `sort.SliceStable` in `sortLogRows` to improve performance
* re-use `resultRows`, `tailRows`, and `perStreamRows` instead of re-allocating them during every iteration
* use `slices.BinarySearchFunc` over linear search to find where to skip to

Fusl commented:
rebased from master and made some small adjustments

valyala commented on app/vlselect/logsql/logsql.go:
```diff
@@ -453,16 +480,25 @@ func ProcessLiveTailRequest(ctx context.Context, w http.ResponseWriter, r *http.
 func sortLogRows(rows []logRow) {
-	sort.SliceStable(rows, func(i, j int) bool {
-		return rows[i].timestamp < rows[j].timestamp
+	slices.SortStableFunc(rows, func(a, b logRow) int {
```
This change substitutes two lines of clear code with 8 lines of not so clear code. While `slices.SortStableFunc()` may work faster than `sort.SliceStable()`, this is irrelevant in this context, since it is expected that `sortLogRows` is used for sorting small number of logs.

Also, performance-optimization pull requests must be accompanied with benchmarks, which clearly show performance gains after applying the given performance optimizations.

valyala commented on app/vlselect/logsql/logsql.go:
```diff
@@ -542,30 +580,47 @@ func (tp *tailProcessor) getTailRows() ([][]logstorage.Field, error) {
 			// Skip already written rows
-			for len(rows) > 0 && rows[0].timestamp <= lastTimestamp {
-				rows = rows[1:]
+			i, _ := slices.BinarySearchFunc(rows, lastTimestamp, func(row logRow, timestamp int64) int {
```
This optimization is irrelevant for this particular case, since it is expected that live tailing returns logs at low rate.

Also, this optimization complicates code significantly comparing to the original code, so it would be great to include benchmarks in this PR, which clearly show the benefit of this optimization.

valyala commented on app/vlselect/logsql/logsql.go:
```diff
@@ -473,6 +509,8 @@ type tailProcessor struct {
 
 	perStreamRows  map[string][]logRow
 	lastTimestamps map[string]int64
+	resultRows     []logRow
```
This optimization complicates the code, while it doesn't bring significant benefits for this particular case with live tailing. So it is unlikely this optimization should be added here.

valyala reviewed (changes requested):
@Fusl , thanks for the pull request!

The `offset` query arg looks useful, while all the optimizations in this pull request look dubious.

Please see also the commit 3a5f1019ba00f722011c2d244d72b53a7b358e6b , which adds `start_offset` query arg for fetching historical logs just before the start of live tailing.

Fusl commented on app/vlselect/logsql/logsql.go:
```diff
@@ -420,9 +428,18 @@ func ProcessLiveTailRequest(ctx context.Context, w http.ResponseWriter, r *http.
+		end = time.Now().UnixNano() - tailOffsetEndNsecs
+
+		// Limit the search window to prevent out of memory situations when we fall too far behind.
+		if end-start > tailMaxWindowNsecs {
```
While I understand that live-tailing is primarily intended for consumption by humans, I'm particularly worried about a situation where a human is consuming the live tail log, then pausing the stream and leaving it paused for a long time. Once the stream is resumed, the next query iteration consumes so much memory that it may cause VictoriaLogs to run out of memory. I consider this a DoS vulnerability even if not intentionally triggered.

valyala replied:
@Fusl , thanks for the explanation! This sounds reasonable, but this protection isn't going to be added because of the following reason.

VictoriaLogs querying API provides countless ways to shoot yourself in the foot and execute a query, which will eat all the available CPU and RAM (aka "DoS vulnerabilities"). For example, an innocent-looking query `* | uniq_values(_time)` may eat all the available memory, since it tries fetching `_time` field across all the logs and then returning them as JSON array via [`uniq_values` stats function](https://docs.victoriametrics.com/victorialogs/logsql/#uniq_values-stats). If the number of logs stored at VictoriaLogs is big (e.g. billions), then this query will definitely consume almost all the available disk read IO, CPU and RAM during the execution. It may not lead to OOM crash, since VicitoriaLogs contains memory usage tracking mechanism, which should automatically cancel the query with the error if it consumes too much memory. But this mechanism is easy to bypass by executing many heavy queries at the same time. The main purpose of this mechanism is to prevent from OOM crashes when users execute some heavy ad-hoc queries. It doesn't prevent from DoS attacks.

The threat model for VictoriaLogs is the following: it is expected that it runs in a protected network without direct access from the Internet. It is expected that access to it is provided via auth proxy such as [vmauth](https://docs.victoriametrics.com/vmauth/). VictoriaLogs doesn't provide protection from heavy queries to the supported [HTTP querying endpoints](https://docs.victoriametrics.com/victorialogs/querying/#http-api), since this is impossible task in general case. It is better to clearly document this (TBD) instead of complicating the code with partially-working protection.

valyala replied:
The security threat model for VictoriaLogs is documented at https://docs.victoriametrics.com/victorialogs/#security

Fusl commented:
@valyala I've rebased onto master again and removed the performance improvements for now. I'll include them in a separate PR. This PR contains only the change to add the `offset` argument.

valyala commented:
@Fusl , thanks for the pull request! I transformed it to the commit a44787372f64daa7e21b7670d6aacf6922be4aa9 . Closing the pull request then.

See also the commit 62e6c9bd6f281708f70c44bf3a2c4cf22e370269 , which adds `Security` chapter to VictoriaLogs docs - see https://docs.victoriametrics.com/victorialogs/#security .

valyala commented:
FYI, the `offset` query arg at [`/select/logsql/tail` endpoint](https://docs.victoriametrics.com/victorialogs/querying/#live-tailing) is available starting from [v0.42.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/).
